nats.go

```diff
@@ -264,7 +264,24 @@ const (
 	nuidSize = 22
 )
 
-// A Conn represents a bare connection to a nats-server.
+// A Publisher publishes a NATS message.
+type Publisher interface {
+	// Publish publishes the data argument to the given subject. The data
+	// argument is left untouched and needs to be correctly interpreted on
+	// the receiver.
+	Publish(string, []byte) error
+
+	// PublishMsg publishes the Msg structure, which includes the
+	// Subject, an optional Reply and an optional Data field.
+	PublishMsg(m *Msg) error
+
+	// PublishRequest will perform a Publish() excpecting a response on the
+	// reply subject. Use Request() for automatically waiting for a response
+	// inline.
+	PublishRequest(subj, reply string, data []byte) error
+}
+
+// A Conn implements Publisher and represents a bare connection to a nats-server.
```

Comment:
`Conn` would implement the `Publisher` interface here, but an interface type is more helpful when declared outside of the package where the implementation exists (so that you can use the behavior of a type without having to import the package)

Reply:
I don't want to abstract from nats and use `Publisher` as abstract publishing to somewhere. I offer `Publisher` as the interface to encapsulate implementation. It can be helpful in case when you want to expand behavior of `Conn`. 

In my point of view, in this case `Publisher` should be in the package.

Reply:
Actually, I do agree with Wally. I am not sure why the interface in nats package would help nats users. If I was working on a project that need to integrate with different messaging vendors, I would probably come up with a common interface, and since likely each vendor would have a different send/publish method, I would have to have encapsulate each vendor and implement the publisher interface. If you could give me a concrete example on how the Publisher interface in nats package would help, that may change my opinion. Thanks!

Reply:
I will try to describe my concrete example in go-kit.
There is http transport in that package and also there is `http.ResponseWriter` interface in stdlib. Client code of go-kit decides to what write by this interface. It means client code mustn't know about the implementation of `http.ResponseWriter`. It gives an ability to go-kit to redefine that writer on its own writer. But client code still works with the interface of the `http` package.
For nats, client code must work with `Conn` struct and go-kit doesn't have the ability to redefine publisher. One of the possible solutions to define the interface in go-kit. But it doesn't release from importing of the `nats`. Also, we can use the `http` package as an example. Because there is `ResponseWriter` interface there. For me, it looks similar.

The difference between two solutions looks like this (it's client code)
`func encodeResponse(_ context.Context, reply string, p kitnatstransport.Publisher, response interface{}) error {`
`func encodeResponse(_ context.Context, reply string, p nats.Publisher, response interface{}) error {`

Client code for http transport
`func encodeResponse(_ context.Context, w http.ResponseWriter, response interface{}) error {`

I understand that you have only one implementation and don't see sense to add interface. But maybe it's a good solution to add some abstract contracts like `Publisher`, `Subscriber` and so on?

Reply:
@kozlovic what do you think? if you don't agree with my changes I will try to solve my case another way.

Reply:
@kirooha Sorry for the delay. After some internal discussions, we still feel that having the interface defined outside of the implementation makes more sense. So at this point we are not going to include this in the package. We do appreciate you using NATS and your contribution and feedback. If we later decide to have such interface, we will reference this PR and reach out to you. Thanks again!